lib/matplotlib/backends/backend_mixed.py

```diff
@@ -84,50 +83,49 @@ def start_rasterizing(self):
+        self._renderer = self._raster_renderer
 
     def stop_rasterizing(self):
         """
         Exit "raster" mode.  All of the drawing that was done since
         the last `start_rasterizing` call will be copied to the
         vector backend by calling draw_image.
-
-        If `start_rasterizing` has been called multiple times,
-        `stop_rasterizing` must be called the same number of times before
-        "raster" mode is exited.
         """
-        self._rasterizing -= 1
-        if self._rasterizing == 0:
-            self._renderer = self._vector_renderer
-
-            height = self._height * self.dpi
-            buffer, bounds = self._raster_renderer.tostring_rgba_minimized()
-            l, b, w, h = bounds
-            if w > 0 and h > 0:
-                image = np.frombuffer(buffer, dtype=np.uint8)
-                image = image.reshape((h, w, 4))
-                image = image[::-1]
-                gc = self._renderer.new_gc()
-                # TODO: If the mixedmode resolution differs from the figure's
-                #       dpi, the image must be scaled (dpi->_figdpi). Not all
-                #       backends support this.
-                self._renderer.draw_image(
-                    gc,
-                    l * self._figdpi / self.dpi,
-                    (height-b-h) * self._figdpi / self.dpi,
-                    image)
-            self._raster_renderer = None
-            self._rasterizing = False
 
-            # restore the figure dpi.
-            self.figure.set_dpi(self._figdpi)
+        self._renderer = self._vector_renderer
+
+        height = self._height * self.dpi
+        buffer, bounds = self._raster_renderer.tostring_rgba_minimized()
+        l, b, w, h = bounds
+        if w > 0 and h > 0:
+            image = np.frombuffer(buffer, dtype=np.uint8)
+            image = image.reshape((h, w, 4))
+            image = image[::-1]
+            gc = self._renderer.new_gc()
+            # TODO: If the mixedmode resolution differs from the figure's
+            #       dpi, the image must be scaled (dpi->_figdpi). Not all
+            #       backends support this.
+            self._renderer.draw_image(
+                gc,
+                l * self._figdpi / self.dpi,
+                (height-b-h) * self._figdpi / self.dpi,
+                image)
+        self._raster_renderer = None
+
+        # restore the figure dpi.
+        self.figure.set_dpi(self._figdpi)
 
         if self._bbox_inches_restore:  # when tight bbox is used
             r = process_figure_for_rasterizing(self.figure,
                                                self._bbox_inches_restore,
                                                self._figdpi)
             self._bbox_inches_restore = r
+
+    def finalize(self):
+        if self._rasterizing:
+            self.stop_rasterizing()
+            self._rasterizing = False
+        self._renderer.finalize()
```

Comment:
does this need to be defined as a noop in the base class?

Reply:
I think not. Only the pdf and svg backends implement it, and call it themselves from their print_xxx method. Previously MixedModeRenderer passed it straight through to its current renderer, so it gave an error if it was still in raster mode.

Reply:
This seems a bit too brittle to me.  I would rather see an extra flag to `allow_rasterization` like `outer_most` or `suppress_merging` or something that ensures `stop_rasterization` is called at the end and then we set that flag on the decorator we we apply to `Figure.draw`.

Reply:
I think this would need to be an additional decorator, because to add an argument to `allow_rasterization` I would need to change how it is used at every draw definition.

```python
def finalize_rasterization(draw):
    @wraps(draw)
    def draw_wrapper(artist, renderer, *args, **kwargs):
        result = draw(artist, renderer, *args, **kwargs)
        if renderer._rasterizing:
            renderer.stop_rasterizing()
            renderer._rasterizing = False
        return result
    return draw_wrapper
```

Reply:
Re-using a pattern we have else where in the code base I think we can add a flag without having to change anything:

```python
def allow_rasterization(func=None, *, finalize=False):
     if func is None:
         return functools.partial(alllow_rasterization, finalize=finalize)
    # current implementation + using finalize

```

but a second decorator would also be reasonale.

Reply:
I think I prefer a second decorator, as `allow_rasterization` now has quite a lot of logic in it.